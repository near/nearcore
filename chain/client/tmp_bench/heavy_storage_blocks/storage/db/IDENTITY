158689be07aae77c-fcb0799e19e2f09e