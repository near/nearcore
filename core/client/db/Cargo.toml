[package]
name = "substrate-client-db"
version = "0.1.0"
authors = ["Parity Technologies <admin@parity.io>"]

[dependencies]
parking_lot = "0.4"
log = "0.4"
kvdb = "0.1"
kvdb-rocksdb = "0.1.3"
hash-db = { git = "https://github.com/paritytech/trie" }
substrate-primitives = { path = "../../primitives" }
sr-primitives = { path = "../../sr-primitives" }
substrate-client = { path = "../../client" }
substrate-state-machine = { path = "../../state-machine" }
parity-codec = "2.1"
parity-codec-derive = "2.1"
substrate-executor = { path = "../../executor" }
substrate-state-db = { path = "../../state-db" }
substrate-trie = { path = "../../trie" }

[dev-dependencies]
kvdb-memorydb = "0.1"
substrate-keyring = { path = "../../keyring" }
substrate-test-client = { path = "../../test-client" }
