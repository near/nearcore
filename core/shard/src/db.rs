use std::collections::HashMap;
use std::sync::Arc;

use parking_lot::RwLock;

use primitives::hash::CryptoHash;
use primitives::types::Transaction;
use SignedShardBlock;
use storage::{self, extend_with_cache};
use storage::read_with_cache;

type H264 = [u8; 33];

pub struct ShardChainDb {
    storage: Arc<storage::Storage>,
    transaction_addresses: RwLock<HashMap<Vec<u8>, TransactionAddress>>,
}

/// Represents index of extra data in database
#[derive(Copy, Debug, Hash, Eq, PartialEq, Clone)]
pub enum ExtrasIndex {
    /// Transaction address index
    TransactionAddress = 0,
}

fn with_index(hash: &CryptoHash, i: ExtrasIndex) -> H264 {
    let mut result = [0; 33];
    result[0] = i as u8;
    result[1..].clone_from_slice(hash.as_ref());
    result
}

/// Represents address of certain transaction within block
#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct TransactionAddress {
    /// Block hash
    pub block_hash: CryptoHash,
    /// Transaction index within the block
    pub index: usize
}

impl ShardChainDb {
    pub fn new(storage: Arc<storage::Storage>) -> Self {
        Self {
            storage,
            transaction_addresses: RwLock::new(HashMap::new()),
        }
    }

    /// Get the address of transaction with given hash.
    pub fn get_transaction_address(&self, hash: &CryptoHash) -> Option<TransactionAddress> {
        let key = with_index(&hash, ExtrasIndex::TransactionAddress);
        read_with_cache(
            &self.storage.clone(),
            storage::COL_EXTRA,
            &self.transaction_addresses,
            &key,
        )
    }

    pub fn update_for_inserted_block(&self, block: &SignedShardBlock) {
        let updates: HashMap<Vec<u8>, TransactionAddress> = block.body.transactions.iter()
            .enumerate()
            .filter_map(|(i, transaction)| {
                match transaction {
                    Transaction::SignedTransaction(t) => {
                        let key = with_index(
                            &t.transaction_hash(),
                            ExtrasIndex::TransactionAddress,
                        );
                        Some((key.to_vec(), TransactionAddress {
                            block_hash: block.hash,
                            index: i,
                        }))
                    },
                    Transaction::Receipt(_) => None,
                }
            })
            .collect();
        extend_with_cache(
            &self.storage.clone(),
            storage::COL_EXTRA,
            &self.transaction_addresses,
            updates,
        );
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use storage::test_utils::create_memory_db;
    use primitives::types::SignedTransaction;

    #[test]
    fn test_get_transaction_address() {
        let db = ShardChainDb::new(Arc::new(create_memory_db()));
        let t = SignedTransaction::empty();
        let transaction = Transaction::SignedTransaction(SignedTransaction::empty());
        let block = SignedShardBlock::new(
            0,
            0,
            CryptoHash::default(),
            CryptoHash::default(),
            vec![transaction],
            vec![],
        );
        db.update_for_inserted_block(&block);
        let address = db.get_transaction_address(&t.transaction_hash());
        let expected = TransactionAddress {
            block_hash: block.hash,
            index: 0
        };
        assert_eq!(address, Some(expected.clone()));

        let cache_key = with_index(
            &t.transaction_hash(),
            ExtrasIndex::TransactionAddress,
        );
        let read = db.transaction_addresses.read();
        let v = read.get(&cache_key.to_vec());
        assert_eq!(v.unwrap(), &expected.clone());
    }
}
