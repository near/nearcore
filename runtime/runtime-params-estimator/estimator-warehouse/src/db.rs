//! Database connection and struct representing rows in the data tables.

use std::path::Path;

use chrono::NaiveDateTime;
use rusqlite::{params, Connection, Row};

use crate::Metric;

/// Wrapper around database connection
pub(crate) struct Db {
    conn: Connection,
}

impl Db {
    pub(crate) fn new(conn: Connection) -> Self {
        Self { conn }
    }

    /// Opens an existing SQLite Db or creates it
    pub(crate) fn open(path: &Path) -> anyhow::Result<Self> {
        let conn = Connection::open(path)?;
        let init_sql = include_str!("init.sql");
        conn.execute_batch(init_sql)?;
        Ok(Self::new(conn))
    }
}

/// A single data row in the estimation table
#[derive(Debug, PartialEq)]
pub(crate) struct EstimationRow {
    /// Name of the estimation / parameter
    pub name: String,
    /// The estimation result converted to gas units
    pub gas: f64,
    /// Parameter for which this estimation is used
    pub parameter: Option<String>,
    /// The estimated time in nanoseconds, (if time-based estimation)
    pub wall_clock_time: Option<f64>,
    /// The number of operations counted (if icount-based estimation)
    pub icount: Option<f64>,
    /// The number of IO read bytes counted (if icount-based estimation)
    pub io_read: Option<f64>,
    /// The number of IO write bytes counted (if icount-based estimation)
    pub io_write: Option<f64>,
    /// For measurements that had some kind of inaccuracies or problems
    pub uncertain_reason: Option<String>,
    /// Which git commit this has been estimated on
    pub commit_hash: String,
}

/// A single data row in the parameter table
#[derive(Debug, PartialEq)]
pub(crate) struct ParameterRow {
    /// Name of the estimation / parameter
    pub name: String,
    /// The estimation result converted to gas units
    pub gas: f64,
    /// Protocol version for which the parameter is valid
    pub protocol_version: u32,
}

impl EstimationRow {
    const SELECT_ALL: &'static str =
        "name,gas,parameter,wall_clock_time,icount,io_read,io_write,uncertain_reason,commit_hash";
    pub fn get(db: &Db, name: &str, commit: &str, metric: Metric) -> anyhow::Result<Vec<Self>> {
        Ok(Self::get_any_metric(db, name, commit)?
            .into_iter()
            .filter(|row| row.is_metric(metric))
            .collect())
    }
    pub fn get_any_metric(db: &Db, name: &str, commit: &str) -> anyhow::Result<Vec<Self>> {
        let select = Self::SELECT_ALL;
        let mut stmt = db.conn.prepare(&format!(
            "SELECT {select} FROM estimation WHERE name = ?1 AND commit_hash = ?2;"
        ))?;
        let data = stmt
            .query_map([name, commit], Self::from_row)?
            .collect::<Result<Vec<_>, rusqlite::Error>>()?;
        Ok(data)
    }
    pub(crate) fn insert(&self, db: &Db) -> anyhow::Result<()> {
        db.conn.execute(
            "INSERT INTO estimation(name,gas,parameter,wall_clock_time,icount,io_read,io_write,uncertain_reason,commit_hash) values (?1,?2,?3,?4,?,?6,?7,?8,?9)",
            params![
                self.name,
                self.gas,
                self.parameter,
                self.wall_clock_time,
                self.icount,
                self.io_read,
                self.io_write,
                self.uncertain_reason,
                self.commit_hash,
            ],
        )?;
        Ok(())
    }
    pub fn select_by_commit_and_metric(
        db: &Db,
        commit: &str,
        metric: Metric,
    ) -> anyhow::Result<Vec<Self>> {
        let select = Self::SELECT_ALL;
        let metric_condition = metric.condition();
        let mut stmt = db.conn.prepare(&format!(
            "SELECT {select} FROM estimation WHERE commit_hash = ?1 AND {metric_condition};"
        ))?;
        let data = stmt
            .query_map([commit], Self::from_row)?
            .collect::<Result<Vec<_>, rusqlite::Error>>()?;
        Ok(data)
    }

    /// Returns one (commit_hash,date) tuple for each commit in store,
    /// optionally filtered by estimation metric. The output is sorted by the
    /// date, in ascending order. Note that the date is not the committed-date
    /// but instead the earliest date associated with an estimation of that
    /// commit, as recorded in the warehouse. This is usually the date the
    /// estimation has been performed.
    pub fn commits_sorted_by_date(
        db: &Db,
        metric: Option<Metric>,
    ) -> anyhow::Result<Vec<(String, NaiveDateTime)>> {
        let metric_condition = match metric {
            Some(m) => format!("WHERE {}", m.condition()),
            None => "".to_owned(),
        };
        let sql = format!("SELECT commit_hash,min(date) FROM estimation {metric_condition} GROUP BY commit_hash ORDER BY date ASC;");
        let mut stmt = db.conn.prepare(&sql)?;
        let data = stmt
            .query_map([], |row| Ok((row.get::<_, String>(0)?, row.get::<_, NaiveDateTime>(1)?)))?
            .collect::<Result<Vec<_>, rusqlite::Error>>()?;
        Ok(data)
    }
    pub fn count_by_metric(db: &Db, metric: Metric) -> anyhow::Result<u64> {
        let sql = match metric {
            Metric::ICount => "SELECT COUNT(*) FROM estimation WHERE icount IS NOT NULL;",
            Metric::Time => "SELECT COUNT(*) FROM estimation WHERE wall_clock_time IS NOT NULL;",
        };
        let count = db.conn.query_row::<i32, _, _>(sql, [], |row| row.get(0))?;
        Ok(count as u64)
    }
    pub fn last_updated(db: &Db, metric: Metric) -> anyhow::Result<Option<NaiveDateTime>> {
        let sql = match metric {
            Metric::ICount => "SELECT MAX(date) FROM estimation WHERE icount IS NOT NULL;",
            Metric::Time => "SELECT MAX(date) FROM estimation WHERE wall_clock_time IS NOT NULL;",
        };
        let dt = db.conn.query_row::<Option<NaiveDateTime>, _, _>(sql, [], |row| row.get(0))?;
        Ok(dt)
    }
    fn is_metric(&self, metric: Metric) -> bool {
        match metric {
            Metric::ICount => self.icount.is_some(),
            Metric::Time => self.icount.is_none() && self.wall_clock_time.is_some(),
        }
    }
    fn from_row(row: &Row) -> rusqlite::Result<Self> {
        Ok(Self {
            name: row.get(0)?,
            gas: row.get(1)?,
            parameter: row.get(2)?,
            wall_clock_time: row.get(3)?,
            icount: row.get(4)?,
            io_read: row.get(5)?,
            io_write: row.get(6)?,
            uncertain_reason: row.get(7)?,
            commit_hash: row.get(8)?,
        })
    }
}

impl ParameterRow {
    pub fn count(db: &Db) -> anyhow::Result<u64> {
        let sql = "SELECT COUNT(*) FROM parameter;";
        let count = db.conn.query_row::<u64, _, _>(sql, [], |row| row.get(0))?;
        Ok(count)
    }
    pub fn latest_protocol_version(db: &Db) -> anyhow::Result<Option<u32>> {
        let sql = "SELECT MAX(protocol_version) FROM parameter;";
        let max = db.conn.query_row::<Option<u32>, _, _>(sql, [], |row| row.get(0))?;
        Ok(max)
    }
}

impl Metric {
    fn condition(&self) -> &'static str {
        match self {
            Metric::ICount => "icount IS NOT NULL",
            Metric::Time => "wall_clock_time IS NOT NULL",
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Db;
    use crate::import::ImportConfig;
    use chrono::{NaiveDate, NaiveDateTime};
    use rusqlite::{functions::FunctionFlags, Connection};

    impl Db {
        /// Create a new in-memory test database with a mocked `datetime()` function.
        pub(crate) fn test() -> Self {
            let conn = Connection::open_in_memory().unwrap();
            let init_sql = include_str!("init.sql");
            conn.execute_batch(init_sql).unwrap();
            let db = Self::new(conn);
            db.mock_time(
                NaiveDate::from_ymd_opt(2015, 5, 15).unwrap().and_hms_opt(11, 22, 33).unwrap(),
            );
            db
        }

        /// Create a new in-memory test database with data defined by the input.
        ///
        /// The test data is expected to come in blocks of JSON lines with a commit header.
        /// WAIT statements can be used as barriers to make sure DB timestamps differ.
        pub(crate) fn test_with_data(input: &str) -> Self {
            let db = Self::test();
            for block in input.split("\n\n") {
                if block.trim() == "WAIT" {
                    // Update mocked time to ensure the following data is considered to be later.
                    let mocked_now: NaiveDateTime =
                        db.conn.query_row("SELECT datetime('now')", [], |r| r.get(0)).unwrap();
                    db.mock_time(mocked_now + chrono::Duration::seconds(1));
                } else {
                    let (commit_hash, input) = block.split_once("\n").unwrap();
                    let conf = ImportConfig {
                        commit_hash: Some(commit_hash.to_string()),
                        protocol_version: None,
                    };
                    db.import_json_lines(&conf, input).unwrap();
                }
            }
            db
        }

        pub(crate) fn mock_time(&self, dt: NaiveDateTime) {
            let string_rep = dt.format("%Y-%m-%d %H:%M:%S").to_string();
            self.conn
                .create_scalar_function("datetime", 1, FunctionFlags::SQLITE_UTF8, move |_ctx| {
                    Ok(string_rep.clone())
                })
                .unwrap();
        }
    }
}
