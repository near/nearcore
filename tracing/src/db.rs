use crate::primitives::RawTrace;
use bson::doc;
use mongodb::{Collection, IndexModel};

/// Simple wrapper on top of mongodb.
#[derive(Clone)]
pub struct Database {
    pub db: mongodb::Database,
}

impl Database {
    /// The collection where we dump all the traces received by the collector.
    pub fn raw_traces(&self) -> Collection<RawTrace> {
        self.db.collection("RawTrace")
    }

    pub async fn new(mongodb_uri: &str, initialize: bool) -> Self {
        let db = mongodb::Client::with_uri_str(mongodb_uri).await.unwrap().database("near");
        let db = Self { db };
        if initialize {
            db.create_indexes().await;
        }
        println!("Mongodb initialized");
        db
    }

    async fn create_indexes(&self) {
        // TODO: add a TTL to the collection so that data is automatically
        // garbage collected by MongoDB.
        let span_chunks = self.raw_traces();
        span_chunks
            .create_index(IndexModel::builder().keys(doc! {"min_time": 1}).build(), None)
            .await
            .unwrap();
        span_chunks
            .create_index(IndexModel::builder().keys(doc! {"max_time": 1}).build(), None)
            .await
            .unwrap();
    }
}
