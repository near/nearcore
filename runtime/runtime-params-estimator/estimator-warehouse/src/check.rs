use crate::db::{Db, EstimationRow};
use crate::zulip::{ZulipEndpoint, ZulipReport};
use crate::Metric;
use std::collections::BTreeSet;

#[derive(clap::Parser, Debug)]
pub(crate) struct CheckConfig {
    /// Send notifications from checks to specified stream.
    /// Notifications are sent iff stream or user is set.
    #[clap(long)]
    zulip_stream: Option<String>,
    /// Send notifications from checks to specified Zulip user ID.
    /// Notifications are sent iff stream or user is set.
    #[clap(long)]
    zulip_user: Option<u64>,
    /// Checks have to be done on one specific metric.
    #[clap(long, value_enum)]
    metric: Metric,
    /// First git commit hash used for comparisons, used as base to calculate
    /// the relative changes. If left unspecified, the two commits that were
    /// inserted most recently are compared.
    #[clap(long)]
    commit_before: Option<String>,
    /// Second git commit hash used for comparisons. If left unspecified, the
    /// two commits that were inserted most recently are compared.
    #[clap(long)]
    commit_after: Option<String>,
    /// Names of estimations that should be checked. Leave empty to perform
    /// comparison on all available estimations.
    #[clap(long)]
    estimations: Vec<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub(crate) enum Status {
    Ok = 0,
    Warn = 1,
    // Critical = 2,
}

#[derive(Debug, PartialEq)]
pub(crate) enum Notice {
    RelativeChange(RelativeChange),
    UncertainChange(UncertainChange),
}

#[derive(Debug, PartialEq)]
pub(crate) struct RelativeChange {
    pub estimation: String,
    pub before: f64,
    pub after: f64,
}

#[derive(Debug, PartialEq)]
pub(crate) struct UncertainChange {
    pub estimation: String,
    pub before: String,
    pub after: String,
}

pub(crate) fn check(db: &Db, config: &CheckConfig) -> anyhow::Result<()> {
    let report = create_report(db, config)?;

    // This is the check command output to observe directly in the terminal.
    for change in report.changes() {
        println!("{change:?}");
    }

    let zulip_receiver = {
        if let Some(user) = config.zulip_user {
            Some(ZulipEndpoint::to_user(user)?)
        } else if let Some(stream) = &config.zulip_stream {
            Some(ZulipEndpoint::to_stream(stream.clone())?)
        } else {
            None
        }
    };

    if let Some(zulip) = zulip_receiver {
        zulip.post(&report)?;
    }
    Ok(())
}

pub(crate) fn create_report(db: &Db, config: &CheckConfig) -> anyhow::Result<ZulipReport> {
    let (commit_after, commit_before) = match (&config.commit_after, &config.commit_before) {
        (Some(a), Some(b)) => (a.clone(), b.clone()),
        (None, None) => {
            let mut commits = EstimationRow::commits_sorted_by_date(db, Some(config.metric))?;
            if commits.len() < 2 {
                anyhow::bail!("need data for at least 2 commits to perform comparison");
            }
            (commits.pop().unwrap().0, commits.pop().unwrap().0)
        }
        _ => anyhow::bail!("you have to either specify both commits for comparison or neither"),
    };
    let estimations = if !config.estimations.is_empty() {
        config.estimations.clone()
    } else {
        let rows_a = EstimationRow::select_by_commit_and_metric(db, &commit_after, config.metric)?;
        let rows_b = EstimationRow::select_by_commit_and_metric(db, &commit_before, config.metric)?;
        let estimations_a = rows_a.into_iter().map(|row| row.name).collect::<BTreeSet<_>>();
        let estimations_b = rows_b.into_iter().map(|row| row.name).collect::<BTreeSet<_>>();
        estimations_a.intersection(&estimations_b).cloned().collect()
    };
    let warnings =
        estimation_changes(db, &estimations, &commit_before, &commit_after, 0.1, config.metric)?;

    let warnings_uncertain = estimation_uncertain_changes(
        db,
        &estimations,
        &commit_before,
        &commit_after,
        config.metric,
    )?;

    let mut report = ZulipReport::new(commit_before, commit_after);
    for warning in warnings {
        report.add(warning, Status::Warn)
    }
    for warning in warnings_uncertain {
        report.add(warning, Status::Warn)
    }
    Ok(report)
}

fn estimation_changes(
    db: &Db,
    estimation_names: &[String],
    commit_before: &str,
    commit_after: &str,
    tolerance: f64,
    metric: Metric,
) -> anyhow::Result<Vec<Notice>> {
    let mut warnings = Vec::new();
    for name in estimation_names {
        let b = &EstimationRow::get(db, name, commit_before, metric)?[0];
        let a = &EstimationRow::get(db, name, commit_after, metric)?[0];
        let rel_change = (b.gas - a.gas).abs() / b.gas;
        if rel_change > tolerance {
            warnings.push(Notice::RelativeChange(RelativeChange {
                estimation: name.clone(),
                before: b.gas,
                after: a.gas,
            }))
        }
    }

    Ok(warnings)
}

fn estimation_uncertain_changes(
    db: &Db,
    estimation_names: &[String],
    commit_before: &str,
    commit_after: &str,
    metric: Metric,
) -> anyhow::Result<Vec<Notice>> {
    let mut warnings = Vec::new();
    for name in estimation_names {
        let b = EstimationRow::get(db, name, commit_before, metric)?.remove(0);
        let a = EstimationRow::get(db, name, commit_after, metric)?.remove(0);
        match (b.uncertain_reason, a.uncertain_reason) {
            (None, None) => continue,
            (Some(uncertain_before), None) => {
                add_warning(&mut warnings, name.clone(), uncertain_before, "None".to_owned())
            }
            (None, Some(uncertain_after)) => {
                add_warning(&mut warnings, name.clone(), "None".to_owned(), uncertain_after)
            }
            (Some(uncertain_before), Some(uncertain_after)) => {
                add_warning(&mut warnings, name.clone(), uncertain_before, uncertain_after);
            }
        }
    }

    Ok(warnings)
}

fn add_warning(warnings: &mut Vec<Notice>, name: String, before: String, after: String) {
    warnings.push(Notice::UncertainChange(UncertainChange { estimation: name, before, after }))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[track_caller]
    fn generate_test_report(input: &str, metric: Metric, estimations: &[&str]) -> ZulipReport {
        let db = Db::test_with_data(input);
        let config = CheckConfig {
            zulip_stream: None,
            zulip_user: None,
            metric,
            commit_before: None,
            commit_after: None,
            estimations: estimations.iter().map(|&s| s.to_owned()).collect(),
        };
        create_report(&db, &config).unwrap()
    }

    #[test]
    fn test_check_command() {
        let input_a = r#"
        0000a
        {"computed_in":{"nanos":800,"secs":44},"name":"LogBase","result":{"gas":1000000000.0,"instructions":8000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":809,"secs":26},"name":"LogByte","result":{"gas":1000000000.0,"instructions":8000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}

        0001a
        {"computed_in":{"nanos":814,"secs":9},"name":"LogBase","result":{"gas":2000000000.0,"instructions":16000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":694,"secs":33},"name":"LogByte","result":{"gas":1002000000.0,"instructions":8016.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}

        0002a
        {"computed_in":{"nanos":331,"secs":24},"name":"LogBase","result":{"gas":3000000000.0,"instructions":24000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":511,"secs":52},"name":"LogByte","result":{"gas":1004000000.0,"instructions":8032.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}

        0003a
        {"computed_in":{"nanos":633,"secs":7},"name":"LogBase","result":{"gas":4000000000.0,"instructions":32000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":173,"secs":2},"name":"LogByte","result":{"gas":1006000000.0,"instructions":8048.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":633,"secs":7},"name":"UncertainTest","result":{"gas":4000000000.0,"instructions":32000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":"NEGATIVE-COST"}}
        {"computed_in":{"nanos":655,"secs":56},"name":"LogByte","result":{"gas":20000000.0,"time_ns":20,"metric":"time","uncertain_reason":null}}

        0004a
        {"computed_in":{"nanos":319,"secs":19},"name":"LogBase","result":{"gas":5000000000.0,"instructions":40000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":527,"secs":15},"name":"LogByte","result":{"gas":1008000000.0,"instructions":8064.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":633,"secs":7},"name":"UncertainTest","result":{"gas":4000000000.0,"instructions":32000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":"HIGH-VARIANCE"}}
        {"computed_in":{"nanos":633,"secs":7},"name":"UncertainTest2","result":{"gas":4000000000.0,"instructions":32000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":661,"secs":11},"name":"LogByte","result":{"gas":15000000.0,"time_ns":15,"metric":"time","uncertain_reason":null}}
        {"computed_in":{"nanos":661,"secs":11},"name":"LogByte","result":{"gas":15000000.0,"time_ns":15,"metric":"time","uncertain_reason":null}}
        {"computed_in":{"nanos":0,"secs":0},"name":"AltBn128Sum","result":{"gas":0.0,"time_ns":0,"metric":"time","uncertain_reason":null}}
        {"computed_in":{"nanos":0,"secs":0},"name":"AltBn128MultiExp","result":{"gas":0.0,"time_ns":0,"metric":"time","uncertain_reason":null}}
        "#;

        // Only "LogBase" changes enough to show up in report.
        let report = generate_test_report(input_a, Metric::ICount, &[]);
        insta::assert_snapshot!(report.to_string());

        // Add more data and verify the notifications are updated.
        let input_b = input_a.to_owned()
            + r#"

        WAIT

        0000b
        {"computed_in":{"nanos":119,"secs":46},"name":"LogBase","result":{"gas":6000000000.0,"instructions":48000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":372,"secs":12},"name":"LogByte","result":{"gas":7000000000.0,"instructions":56000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":633,"secs":7},"name":"UncertainTest","result":{"gas":4000000000.0,"instructions":32000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":633,"secs":7},"name":"UncertainTest2","result":{"gas":4000000000.0,"instructions":32000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":"BLOCK-MEASUREMENT-OVERHEAD"}}
        {"computed_in":{"nanos":262,"secs":15},"name":"LogByte","result":{"gas":20000000.0,"time_ns":20,"metric":"time","uncertain_reason":null}}
        {"computed_in":{"nanos":0,"secs":0},"name":"AltBn128Sum","result":{"gas":0.0,"time_ns":0,"metric":"time","uncertain_reason":null}}
        {"computed_in":{"nanos":0,"secs":0},"name":"AltBn128MultiExp","result":{"gas":10.0,"time_ns":10,"metric":"time","uncertain_reason":null}}
        "#;

        // Now both estimations have changed.
        let report = generate_test_report(&input_b, Metric::ICount, &[]);
        insta::assert_snapshot!(report.to_string());

        // Verify that filter for specific estimations works.
        let report = generate_test_report(&input_b, Metric::ICount, &["LogBase"]);
        insta::assert_snapshot!(report.to_string());

        // Filter for metric.
        let report = generate_test_report(&input_b, Metric::Time, &[]);
        insta::assert_snapshot!(report.to_string());
    }
}
