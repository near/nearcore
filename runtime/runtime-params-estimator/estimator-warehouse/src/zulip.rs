use std::env;

use anyhow::Context;
use reqwest::blocking::Client;

use crate::check::{Notice, RelativeChange, Status, UncertainChange};

const ZULIP_SERVER: &str = "near.zulipchat.com";

pub(crate) struct ZulipEndpoint {
    client: Client,
    full_endpoint_url: String,
    user_list: Option<String>,
    stream: Option<String>,
}

pub(crate) struct ZulipReport {
    status: Status,
    before: String,
    after: String,
    changes: Vec<RelativeChange>,
    changes_uncertain: Vec<UncertainChange>,
}

impl ZulipEndpoint {
    pub(crate) fn to_user(user: u64) -> anyhow::Result<Self> {
        Ok(Self {
            client: Client::new(),
            full_endpoint_url: Self::form_url()?,
            stream: None,
            user_list: Some(format!("[{user}]")),
        })
    }
    pub(crate) fn to_stream(stream: String) -> anyhow::Result<Self> {
        Ok(Self {
            client: Client::new(),
            full_endpoint_url: Self::form_url()?,
            stream: Some(stream),
            user_list: None,
        })
    }
    pub(crate) fn post(&self, report: &ZulipReport) -> anyhow::Result<()> {
        self.send_raw_message(&report.to_string(), "Bot reports")
    }
    fn form_url() -> anyhow::Result<String> {
        let bot_email =
            env::var("ZULIP_BOT_EMAIL").context("ZULIP_BOT_EMAIL environment variable not set")?;
        let api_key = env::var("ZULIP_BOT_API_KEY")
            .context("ZULIP_BOT_API_KEY environment variable not set")?;
        Ok(format!("https://{bot_email}:{api_key}@{ZULIP_SERVER}/api/v1/messages"))
    }
    fn send_raw_message(&self, msg: &str, topic: &str) -> anyhow::Result<()> {
        let params = if let Some(user_list) = &self.user_list {
            vec![("type", "private"), ("to", user_list), ("content", msg)]
        } else {
            vec![
                ("type", "stream"),
                ("to", self.stream.as_deref().unwrap()),
                ("topic", topic),
                ("content", msg),
            ]
        };
        self.client.post(&self.full_endpoint_url).form(&params).send()?;
        Ok(())
    }
}

impl ZulipReport {
    pub(crate) fn new(before: String, after: String) -> Self {
        Self { status: Status::Ok, before, after, changes: vec![], changes_uncertain: vec![] }
    }
    pub(crate) fn add(&mut self, warning: Notice, status: Status) {
        self.status = std::cmp::max(self.status, status);
        match warning {
            Notice::RelativeChange(change) => self.changes.push(change),
            Notice::UncertainChange(change) => self.changes_uncertain.push(change),
        }
    }

    pub(crate) fn changes(&self) -> &[RelativeChange] {
        self.changes.as_ref()
    }
}

impl std::fmt::Display for ZulipReport {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        writeln!(f, "## Report ")?;
        writeln!(f, "*Status: {:?}*", self.status)?;
        writeln!(f, "*Current commit: {}*", self.after)?;
        writeln!(f, "*Compared to: {}*", self.before)?;
        writeln!(f, "### Relative gas estimation changes above threshold: {}", self.changes.len())?;
        if !self.changes.is_empty() {
            writeln!(f, "```")?;
            for change in &self.changes {
                let percent_change = 100.0 * (change.after - change.before) / change.before;
                writeln!(
                    f,
                    "{:<40} {:>16} ➜ {:>16} ({}{:.2}%)",
                    change.estimation,
                    format_gas(change.before),
                    format_gas(change.after),
                    if percent_change >= 0.0 { "+" } else { "" },
                    percent_change,
                )?;
            }
            writeln!(f, "```")?;
        }
        writeln!(f, "### Gas estimator uncertain estimations: {}", self.changes_uncertain.len())?;
        if !self.changes_uncertain.is_empty() {
            writeln!(f, "```")?;
            for change in &self.changes_uncertain {
                writeln!(
                    f,
                    "{:<40} {:>32} ➜ {:<32}",
                    change.estimation, change.before, change.after,
                )?;
            }
            writeln!(f, "```")?;
        }
        Ok(())
    }
}

fn format_gas(gas: f64) -> String {
    match gas {
        n if n > 1e12 => format!("{:.2} Tgas", n / 1e12),
        n if n > 1e9 => format!("{:.2} Ggas", n / 1e9),
        n if n > 1e6 => format!("{:.2} Mgas", n / 1e6),
        n => format!("{:.0} gas", n),
    }
}

#[test]
fn test_format_gas() {
    assert_eq!(format_gas(0.0).as_str(), "0 gas");
    assert_eq!(format_gas(12345.0).as_str(), "12345 gas");
    assert_eq!(format_gas(123e6).as_str(), "123.00 Mgas");
    assert_eq!(format_gas(123.456e9).as_str(), "123.46 Ggas");
    assert_eq!(format_gas(0.456e12).as_str(), "456.00 Ggas");
    assert_eq!(format_gas(123.456e12).as_str(), "123.46 Tgas");
    assert_eq!(format_gas(123.456e15).as_str(), "123456.00 Tgas");
}
