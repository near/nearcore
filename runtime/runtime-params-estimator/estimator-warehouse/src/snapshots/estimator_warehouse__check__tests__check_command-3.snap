---
source: runtime/runtime-params-estimator/estimator-warehouse/src/check.rs
expression: report.to_string()
---
## Report 
*Status: Warn*
*Current commit:         0000b*
*Compared to:         0004a*
### Relative gas estimation changes above threshold: 1
```
LogBase                                         5.00 Ggas âžœ        6.00 Ggas (+20.00%)
```
### Gas estimator uncertain estimations: 0

